File: frontend/src/api/group.ts
import client from "./client";

// ── Factory CRUD ──

export interface Factory {
  id: string;
  code: string;
  name: string;
  location?: string;
  is_active: boolean;
}

export interface FactoryListResponse {
  items: Factory[];
  total: number;
}

export const listFactories = (isActive?: boolean) =>
  client.get<FactoryListResponse>("/group/factories", {
    params: { is_active: isActive },
  });

export const createFactory = (data: {
  code: string;
  name: string;
  location?: string;
}) => client.post<Factory>("/group/factories", data);

export const updateFactory = (
  factoryId: string,
  data: { name?: string; location?: string; is_active?: boolean }
) => client.put<Factory>(`/group/factories/${factoryId}`, data);

export const deactivateFactory = (factoryId: string) =>
  client.delete<void>(`/group/factories/${factoryId}`);

// ── Group Dashboard ──

export interface FactoryKPI {
  factory_id: string;
  factory_code: string;
  factory_name: string;
  open_fmea_count: number;
  open_capa_count: number;
  overdue_capa_count: number;
  active_spc_alarms: number;
  pending_iqc_inspections: number;
  open_scars: number;
  open_supplier_risk_alerts: number;
  recent_audit_findings: number;
}

export interface GroupDashboardResponse {
  factories: FactoryKPI[];
  totals: FactoryKPI;
  snapshot_date: string | null;
}

export const getGroupDashboard = () =>
  client.get<GroupDashboardResponse>("/group/dashboard");

// ── Factory Comparison ──

export interface FactoryComparisonItem {
  factory_id: string;
  factory_code: string;
  factory_name: string;
  metrics: Record<string, number>;
}

export interface FactoryComparisonResponse {
  factories: FactoryComparisonItem[];
  metric_names: string[];
}

export const getFactoryComparison = (metricNames?: string[]) =>
  client.get<FactoryComparisonResponse>("/group/comparison", {
    params: metricNames ? { metric_names: metricNames.join(",") } : undefined,
  });

// ── Shared Suppliers ──

export interface SharedSupplierResponse {
  shared_profile_id: string | null;
  unified_credit_code: string | null;
  name: string;
  short_name: string | null;
  industry: string | null;
  factory_evaluations: {
    factory_id: string;
    factory_code: string;
    grade: string;
    total_score: number;
  }[];
}

export const getSharedSuppliers = () =>
  client.get<SharedSupplierResponse[]>("/group/suppliers");

export const mergeSuppliers = (data: {
  supplier_ids: string[];
  shared_profile_id?: string;
}) => client.post("/group/suppliers/merge", data);

// ── Cross-Factory Audits ──

export interface CrossFactoryAuditResponse {
  program_id: string;
  program_no: string;
  audit_type: string;
  status: string;
  target_factory_ids: string[];
  target_factory_codes: string[];
  finding_count: number;
}

export const getCrossFactoryAudits = () =>
  client.get<CrossFactoryAuditResponse[]>("/group/audits");

export const getAuditProgramFactories = (programId: string) =>
  client.get<{ factory_id: string; factory_code: string }[]>(
    `/group/audits/${programId}/factories`
  );

export const addFactoryToAuditProgram = (
  programId: string,
  factoryId: string
) =>
  client.post(`/group/audits/${programId}/factories`, { factory_id: factoryId });

export const removeFactoryFromAuditProgram = (
  programId: string,
  factoryId: string
) => client.delete(`/group/audits/${programId}/factories/${factoryId}`);

File: backend/app/schemas/group.py
"""Group API schemas."""
import uuid
from datetime import date

from pydantic import BaseModel, Field


class FactoryKPI(BaseModel):
    """KPI data for a single factory."""
    factory_id: uuid.UUID
    factory_code: str
    factory_name: str
    open_fmea_count: int = 0
    open_capa_count: int = 0
    overdue_capa_count: int = 0
    active_spc_alarms: int = 0
    pending_iqc_inspections: int = 0
    open_scars: int = 0
    open_supplier_risk_alerts: int = 0
    recent_audit_findings: int = 0


class GroupDashboardResponse(BaseModel):
    """Group dashboard aggregated data."""
    factories: list[FactoryKPI]
    totals: FactoryKPI  # Aggregated totals across all factories
    snapshot_date: date | None = None


class FactoryComparisonItem(BaseModel):
    """Comparison data for a single factory."""
    factory_id: uuid.UUID
    factory_code: str
    factory_name: str
    metrics: dict  # Flexible metric key-value pairs


class FactoryComparisonResponse(BaseModel):
    """Factory comparison data."""
    factories: list[FactoryComparisonItem]
    metric_names: list[str]


class SharedSupplierResponse(BaseModel):
    """Shared supplier across factories."""
    shared_profile_id: uuid.UUID | None
    unified_credit_code: str | None
    name: str
    short_name: str | None
    industry: str | None
    factory_evaluations: list[dict]  # [{"factory_id": ..., "factory_code": ..., "grade": ..., "total_score": ...}]


class CrossFactoryAuditResponse(BaseModel):
    """Cross-factory audit program."""
    program_id: uuid.UUID
    program_no: str
    audit_type: str
    status: str
    target_factory_ids: list[uuid.UUID]
    target_factory_codes: list[str] = Field(default_factory=list)
    finding_count: int = 0


class SupplierMergeRequest(BaseModel):
    """Merge two+ supplier records from different factories into a shared profile."""
    supplier_ids: list[uuid.UUID]  # 2+ supplier IDs to merge
    shared_profile_id: uuid.UUID | None = None  # optional, create if None


class MergedSupplierResponse(BaseModel):
    """Result of a supplier merge."""
    shared_profile_id: uuid.UUID
    unified_credit_code: str | None
    name: str
    short_name: str | None
    industry: str | None
    supplier_ids: list[uuid.UUID]
    factory_ids: list[uuid.UUID]


class AuditFactoryAssignment(BaseModel):
    """Assign a factory to an audit program."""
    factory_id: uuid.UUID


class AuditProgramFactoriesResponse(BaseModel):
    """Target factories for an audit program."""
    program_id: uuid.UUID
    factory_ids: list[uuid.UUID]
    factory_codes: list[str] = Field(default_factory=list)